test/e2e/framework/service/jig.go

```diff
@@ -55,6 +56,12 @@ import (
 // NodePortRange should match whatever the default/configured range is
 var NodePortRange = utilnet.PortRange{Base: 30000, Size: 2768}
 
+var staticPortRange = utilnet.PortRange{Base: 30000, Size: 87}
+
+var staticPorts = makeStaticPorts()
+
+var staticPortLock = sync.Mutex{}
+
```

Comment:
This can be simpler, we just need an array with a mutex so create a new type
```
type staticPortRange struct {
         mu sync.Mutex
         ports []bool
}
```
the index of array is the offset of the port, the value the state

so port 30010, will be 30010 minus NodePortRange base = 10
ports[10]

dose it make sense?

Reply:
Thanks for your comment @aojea . Do you suggest to wrap mutex, array , start of port range and count in a struct and initialize that ? I can do that .

The offset + base logic is already being used here and at other place https://github.com/kubernetes/kubernetes/blob/18c8b3423995494ec3f71bf0488652bab7c7dbcd/test/e2e/framework/service/jig.go#L109

Or have I incorrectly implemented it?

Reply:
to my test the functionality is scattered across multiple objects
```
var staticPortRange = utilnet.PortRange{Base: 30000, Size: 87}

var staticPorts = makeStaticPorts()

var staticPortLock = sync.Mutex{}
```

3 new internal variables + 2 new Methods in TestJig.

I always prefer to have better encapsulation, one new type with its own methods, this type should not be exported to avoid people using it, so it can only used via TestJig. The methods in Testjig should also implement the retry functionality like https://github.com/kubernetes/kubernetes/pull/127153/files#r1750021612

This code is not tested just as illustration

```
type staticPortRange struct {
         mu sync.Mutex
         nodeports [87]bool   // this is the static port range based on current NodePortRange KEP-3668
}


// allocate tries to allocate the passed node port and return true if succeeds or false if can not be allocated
func (s *staticPortRange) allocate(port int) bool {
   s.mu.Lock()
  defer s.mu.Unlock()
  // remove the base
  port = port -  NodePortRange.Base 
  // port is out of range, it can not be allocated
  if port < 0 || port > len(s.nodeports) {
     return false
  }
  // port already allocated
  if nodeports[port] {
      return false
  }
  nodeports[port] = true
  return true
}

// allocate next allocates a free port from the range and returns its number
func (s *staticPortRange) allocateNext() (int, bool) {
   s.mu.Lock()
  defer s.mu.Unlock()
 // start in a random offset
  start := rand.Intn(len(s.staticPortRange))
  for i:= 0; i < len(s.staticPortRange); i ++
       port := (start + i) % len(.staticPortRange)
       if s.allocate(port) {
            return NodePortRange.Base + port, true
       }
   }
  return false 
  }

/ release the port passed as argument
func (s *staticPortRange) release(port int) {
   s.mu.Lock()
  defer s.mu.Unlock()
  port = port -  NodePortRange.Base 
  // port is out of range, it can not be allocated
  if port < 0 || port > len(s.nodeports) {
     return
  }
  nodeports[port] = false
}

var(
   staticPortAllocator = &staticPortRange{
       nodeports: ([87]bool{}
       baseport 
   }
)

```

The the TestJig methods just call directly the staticPortAllocator methods

```
func (j *TestJig) GetUnusedStaticNodePortAndReserve() int {
    port, ok := staticPortAllocator.AllocateNext()
}
```